oncall/update.py

```diff
@@ -0,0 +1,203 @@
+        switch_date = previous + datetime.timedelta(days=day_shift)
+        switch_date = switch_date.replace(hour=switch.hour, minute=switch.minute)
+        next_date = min(next_date, switch_date)
+    return next_date
+
+
+def calculate_new_shift_dates(last_shift, extend_upto, shift_switch_dates):
+    new_shift_dates = []
+    while last_shift <= extend_upto:
+        last_shift = next_shift_change(last_shift, shift_switch_dates)
+        new_shift_dates.append(last_shift)
+    return new_shift_dates
+
+
+def optimize_future_shifts(new_shift_dates, past_shift_assignees, victims):
+    possible_schedules = []
+    comb_format = '{:0%db}' % len(new_shift_dates)
+    for victim in victims:
+        past_shifts = [int(victim in a) for a in past_shift_assignees]
+        for n in xrange(2 ** len(new_shift_dates)):
+            possible_schedules.append(
+                (victim, tuple(past_shifts + map(int, comb_format.format(n)))))
+
+    model = pulp.LpProblem("Shift Scheduling Model", pulp.LpMinimize)
+    # pylint: disable=invalid-name
+    x = pulp.LpVariable.dicts('shift', possible_schedules, 0, 1, cat=pulp.LpInteger)
+
+    model += sum([score(sched, len(victims)) * x[(victim, sched)]
+                  for victim, sched in possible_schedules])
+
+    # each person must have one shift schedule
+    for victim in victims:
+        model += sum([x[sched] for sched in possible_schedules
+                      if victim == sched[0]]) == 1, "Unique shifts for %s" % victim
+
+    # only one person is on for each shift
+    for n in xrange(len(new_shift_dates)):
+        model += sum([x[sched] for sched in possible_schedules
+                      if sched[1][len(past_shift_assignees) + n]]) == 1,\
+                 "One person on for future #%d" % n
+
+    model.solve()
+
+    chosen = [[] for _ in xrange(len(new_shift_dates))]
+    for sched in possible_schedules:
+        if x[sched].value() == 1:
+            for n in xrange(len(new_shift_dates)):
+                if sched[1][n + len(past_shift_assignees)]:
+                    chosen[n].append(sched[0])
+
+    return zip(new_shift_dates, chosen)
+
+
+def update(data):
+    options = data['options']
+    shift_switch_dates = [
+        datetime.datetime.strptime(s, SHIFT_FORMAT)
+        for s in options['shift_switch_dates'].split(',')]
+
+    extend_upto = options.get('extend_upto', '6w')
+    if extend_upto.endswith('w'):
+        extend_upto = datetime.timedelta(weeks=int(extend_upto[:-1]))
+    extend_upto += datetime.datetime.now()
+
+    shifts = list(data['shifts'])  # copy
+    last_shift = shifts[-1][0] if shifts else datetime.datetime.now()
+    new_shift_dates = calculate_new_shift_dates(last_shift, extend_upto, shift_switch_dates)
+
+    if len(new_shift_dates) < options.get('extend_at_least', 1):
+        return data  # bail early if there's no work to do
+
+    plan = optimize_future_shifts(new_shift_dates, [s[1] for s in shifts[-16:]], options['victims'])
```

Comment:
`[s[1] for s in shifts[-16:]]` is pretty magical and not at all obvious what it's doing.